R/data.table.R

```diff
@@ -1861,90 +1861,243 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
+  # not change the input data.table. New copies in memory are only 
+  # created for columns that are modified (those coerced to a different
+  # type or class)
+  X = x
+  dm = dim(x)
+  cn = names(x)
+  class(X) = NULL
+  p = length(cn)
+  n = dm[1L]
+  
+  # If there is a rownames column, exctract, drop, and update the dimensions
   if (!is.null(rownames)) {
-    # extract that column and drop it.
-    rownames.value = x[[rownames]]
-    dm = dim(x) - 0:1
-    cn = names(x)[-rownames]
-    X = x[, .SD, .SDcols = cn]
-  } else {
-    dm = dim(x)
-    cn = names(x)
-    X = x
-  }
+    X[[rownames.index]] = NULL
+    dm = dm - 0:1
+    cn = cn[-rownames.index]
+    p = p - 1L
+  }
+  
+  # The maximum dimension size (row or column) for a matrix is 2^31-1 (or the Machine maximum integer)
+  # Check and error before doing computation/memory expensive column checks and coercion
+  if (p > .Machine$integer.max) 
+    error("Matrices with > ", .Machine$integer.max, " (.Machine$integer.max) columns are not supported")
+  if (n > INT_MAX) 
+    error("Matrices with > ", .Machine$integer.max, " (.Machine$integer.max) rows are not supported")
+  
+  # Check for any wide matrix like columns and unpack using as.data.table
+  if (length(which_wide_columns(x)) > 0L) {
+    X = as.data.table(X)
+    dm = dim(X)
+    cn = names(X)
+    class(X) = NULL
+    p = length(cn)
+    n = dm[1L]
+  }
+  
+  # If no rows or columns can simply return empty array
   if (any(dm == 0L))
     return(array(NA, dim = dm, dimnames = list(rownames.value, cn)))
-  p = dm[2L]
-  n = dm[1L]
-  collabs = as.list(cn)
-  class(X) = NULL
-  non.numeric = non.atomic = FALSE
-  all.logical = TRUE
-  for (j in seq_len(p)) {
-    if (is.ff(X[[j]])) X[[j]] = X[[j]][]   # nocov to bring the ff into memory, since we need to create a matrix in memory
-    xj = X[[j]]
-    if (length(dj <- dim(xj)) == 2L && dj[2L] > 1L) {
-      if (inherits(xj, "data.table"))
-        xj = X[[j]] = as.matrix(X[[j]])
-      dnj = dimnames(xj)[[2L]]
-      collabs[[j]] = paste(collabs[[j]], if (length(dnj) >
-        0L)
-        dnj
-      else seq_len(dj[2L]), sep = ".")
-    }
-    if (!is.logical(xj))
-      all.logical = FALSE
-    if (length(levels(xj)) > 0L || !(is.numeric(xj) || is.complex(xj) || is.logical(xj)) ||
-        (!is.null(cl <- attr(xj, "class", exact=TRUE)) && any(cl %chin%
-        c("Date", "POSIXct", "POSIXlt"))))
-      non.numeric = TRUE
-    if (!is.atomic(xj))
-      non.atomic = TRUE
-  }
-  if (non.atomic) {
+  
+  # We now need to make sure all the columns are the same type, handling special
+  # cases where some columns are not stored in memory (ff objects), lists, or 
+  # otherwise non-atomic types (e.g. S4). Further, we want to minimize
+  # the number of checks done, e.g. if all columns are already the same type
+
+  col.classes = lapply(X, class)
+  uniq.col.classes = unique(col.classes) # list structure preserves cases where a column has multiple classes
```

Comment:
I think `unique.list` is substantially slower than `unique` on atomic, so we could do `col.classes = sapply(x, class)` which would return `list` if any non-scalar classes present. Caveat to deal with the `length(x)==1L` case which behaves a bit strangely forcing `matrix` when `length(class(x[[1L]]))>1L`

```
v = sample(c('numeric', 'character', 'logical', 'integer'), 1e6, TRUE)
l = as.list(v)

microbenchmark::microbenchmark(
  times = 1e2,
  unique(v),
  unique(l)
)
# Unit: milliseconds
#       expr      min       lq     mean   median       uq       max neval cld
#  unique(v) 11.31295 13.19363 18.20252 15.88323 19.10761 212.27103   100  a 
#  unique(l) 29.60442 33.22747 36.82811 36.33566 40.50515  46.90842   100   b
```

Reply:
Unfortunately sapply causes problems - it will return a matrix all columns have the same number of classes and that N > 1. 

E.g. test 2136.3 when both columns have class == c("IDate", "Date") then sapply will return a matrix:

```
    A       B      
[1,] "IDate" "IDate"
[2,] "Date"  "Date" 
```